olp-cpp-sdk-authentication/src/AuthenticationClient.cpp

```diff
@@ -466,78 +501,127 @@ AuthenticationClient::Impl::Impl(AuthenticationSettings settings)
-      return {{static_cast<int>(http::ErrorCode::IO_ERROR),
-               "Cannot sign in while offline"}};
-    }
+olp::client::HttpResponse AuthenticationClient::Impl::CallAuth(
+    const client::OlpClient& client, client::CancellationContext context,
+    const AuthenticationCredentials& credentials,
+    const SignInProperties& properties, std::time_t timestamp) {
+  const auto url = settings_.token_endpoint_url + kOauthEndpoint;
+
+  client::OlpClient::ParametersType headers = {
+      {http::kAuthorizationHeader,
+       GenerateHeader(credentials, url, timestamp)}};
+  if (context.IsCancelled()) {
+    return {static_cast<int>(olp::http::ErrorCode::CANCELLED_ERROR),
+            "Sign was cancelled"};
+  }
+  auto auth_response =
+      client.CallApi(kOauthEndpoint, "POST", {}, std::move(headers), {},
+                     GenerateClientBody(properties), kApplicationJson, context);
+  return auth_response;
+}
 
-    client::OlpClient client = CreateOlpClient(settings_, boost::none);
+SignInResult AuthenticationClient::Impl::ParseAuthResponse(
+    int status, std::stringstream& auth_response) {
+  auto document = std::make_shared<rapidjson::Document>();
+  rapidjson::IStreamWrapper stream(auth_response);
+  document->ParseStream(stream);
+  auto resp_impl = std::make_shared<SignInResultImpl>(
+      status, olp::http::HttpErrorToString(status), document);
+  SignInResult response(resp_impl);
 
-    std::time_t timestamp =
-        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+  return response;
+}
 
-    if (!settings_.use_system_time) {
-      auto server_time = GetTimeFromServer(context, client);
-      if (server_time.IsSuccessful()) {
-        timestamp = server_time.GetResult();
-      }
-    }
+AuthenticationClient::SignInClientResponse
+AuthenticationClient::Impl::FindSingInResponceInCache(
+    int status, const std::string& error_message,
+    AuthenticationCredentials credentials) {
+  if (status == static_cast<int>(olp::http::ErrorCode::CANCELLED_ERROR)) {
+    return {{status, error_message}};
+  }
 
-    const auto url = settings_.token_endpoint_url + kOauthEndpoint;
+  auto cached_response = client_token_cache_->locked(
+      [&](utils::LruCache<std::string, SignInResult>& c)
+          -> boost::optional<SignInResult> {
+        auto it = c.Find(credentials.GetKey());
+        return it != c.end() ? boost::make_optional(it->value()) : boost::none;
+      });
 
-    client::OlpClient::ParametersType headers = {
-        {http::kAuthorizationHeader,
-         GenerateHeader(credentials, url, timestamp)}};
+  if (!cached_response) {
+    return {{status, error_message}};
+  }
 
-    auto auth_response = client.CallApi(
-        kOauthEndpoint, "POST", {}, std::move(headers), {},
-        GenerateClientBody(properties), kApplicationJson, context);
+  return *cached_response;
+}
 
-    const auto status = auth_response.status;
+client::CancellationToken AuthenticationClient::Impl::SignInClient(
```

Comment:
Maybe I am missing this but I think you handle cancellation a little wrong. At least you check cancellation status in CallAuth but you do not stop afterwards but keep on going and even trying to parse the response from CallAuth() which in this case is `{static_cast<int>(olp::http::ErrorCode::CANCELLED_ERROR), "Cancelled"}`.
I would recommend that you stop directly and return status to the user that it was cancelled. This means checking after each major step that the result is not cancelled.

Reply:
I have added additional check for context cancellation. Cancelation after CallAuth should be handled in FindSingInResponseInCache.
```
          if (status < 0) {
            return FindSingInResponseInCache(
                status, auth_response.response.str(), credentials);
          } 
```

Reply:
Ok, now I see. I didn't see that you return http:ErrorCode which is negative. Then ignore this comment.